### PR TITLE
Limit `listRoles` API visibility

### PR DESCRIPTION
**Pull request by hsato03**
### Description

When calling the `listRoles` API, users can see roles with more permissions than theirs.

Therefore, the behavior of the `listRoles` API was changed so that users can only see roles that their role has permission to access (roles with same and less permissions).

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

1. I created a custom role based on User role and added the `listRoles` API to it.
2. I created an account using the role from step 1 and logged into it.
3. I called the `listRoles` API via CloudMonkey and verfied that the roles with more permissions than mine were not listed, such as default admin roles.

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
clgtm and unit-test included, but i think some monkey testing is still in order

File: server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java [hunk 1]
```diff
@@ -382,42 +385,100 @@ public List<Role> findRolesByName(String name) {
...
         return count;
     }
 
+    /**
+     * Checks if the role of the caller account has compatible permissions of the specified role.
+     * For each permission of the role of the caller, the roleToAccess needs to contain the same permission.
+     *
+     * @param rolePermissions the permissions of the caller role.
+     * @param roleToAccess the role that the caller role wants to access.
+     * @return True if the role can be accessed with the given permissions; false otherwise.
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java:427]: @hsato03, what do you think about naming these parameters as `sourceRolePermissions` and `targetRole`? IMO, it seems more intuitive. What do you think?

Comment (hsato03) [reply server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java]: Yes, I agree.

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

**Comment by JoaoJandre** [comment]
@blueorangutan package

**Comment by JoaoJandre** [comment]
@DaanHoogland @sureshanaparti @rohityadavcloud @shwstppr could we run the CI here?

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by bernardodemarco — APPROVED** [review]
LGTM, I manually tested the PR in a local environment.

1. I created a role called `list-roles-test` based on the `User` role.
2. I added the `listRoles` API to the created role.
3. I created an account called `list-roles-test-account` with the custom role.
4. Through CloudMonkey, I created a profile for the account and I executed the `sync` command.
5. I called the `listRoles` API and I got the following output:
```bash
(test-roles-api) 🐱 > listRoles
{
  "count": 2,
  "role": [
    {
      "description": "Default user role",
      "id": "8cead084-771e-11ee-8e59-5254003754dc",
      "isdefault": true,
      "ispublic": true,
      "name": "User",
      "type": "User"
    },
    {
      "id": "ca565c1b-8589-4a78-b7eb-531d5505cadd",
      "isdefault": false,
      "ispublic": true,
      "name": "list-roles-test",
      "type": "User"
    }
  ]
}
```
6. As it can be noticed, default admin roles and the ones with more permission than the `list-roles-test` role do not appear in the API's response.

**Pull request merged**